software/speller/data_collection_platform/dcp/api.py

```diff
@@ -38,38 +39,128 @@ def index():
     # TODO return keyboard layout
     return ' '
 
-
-@bp.route('/char_pressed', methods=['POST', 'GET'])
-def char_pressed():
-    if request.method == 'POST':
-        user_char_pressed = request.form['char_pressed']
-
-
 @bp.route("/openbci/start", methods=['POST'])
 def openbci_start():
     # use a separate process to stream BCI data
     from dcp.bci.stream import stream_bci_api
     from multiprocessing import Process
     with current_app.app_context():
-        p = Process(target=stream_bci_api, args=(shared.bci_processes_states,))
+        # Although the subprocess gets initialized inside this scope, it will persist
+        # bci_processes_states references this subprocess_dict
+        subprocess_dict = shared.manager.dict({
+            'character': None,
+            'phase': None,
+            'frequency': None,
+            'q': deque(),
+            'config_id': None, 
+            'bci_config': None,
+            'state': 'start'
+        })
+        # only pass this subprocess dict to ensure that locks are not too contentious
+        p = Process(target=stream_bci_api, args=(subprocess_dict,))
+        # need to start process before referencing it to obtain the right process_id
         p.start()
-        shared.bci_processes_states[p.pid] = True
+        shared.bci_processes_states[p.pid] = subprocess_dict
+    while subprocess_dict['state'] != 'ready' or subprocess_dict['bci_config'] == None:
+        print('BCI NOT READY YET')
+        time.sleep(1)
+    config = OpenBCIConfig(configuration=subprocess_dict['bci_config'])
+    db.session.add(config)
+    db.session.commit()
+    subprocess_dict['config_id'] = config.id
+    # once the subprocess is ready, return from call
     return {"data": {"pid": p.pid}}, 201
 
 
-@bp.route('/openbci/<int:process_id>/ready', methods=['GET'])
-def openbci_ready(process_id: int):
-    # NOTE: `pid` is currently ignored, see TODO in openbci_stop.
-    with shared.is_bci_ready.get_lock():
-        return {"data": bool(shared.is_bci_ready.value)}, 200
-
+@bp.route('/openbci/<int:process_id>/collect/start', methods=['POST'])
+def openbci_process_collect_start(process_id:int):
+    data = request.form
+    expected_keys = ['character', 'phase' ,'frequency']
+    for key in expected_keys:
+        if key not in request.form.keys():
+            return {'Invalid Request': 'Did not contain all attributes, make sure you send character, phase and frequency'}, 400
+    # We now know that the request contains all the keys
+    if process_id not in shared.bci_processes_states:
+        return {'Invalid Request': 'There is no process with this id, make sure your process id is valid'}, 404
+    subprocess_dict = shared.bci_processes_states[process_id]
+    try:
+        subprocess_dict['character'] = data.get('character')
+        subprocess_dict['frequency'] = float(data.get('frequency'))
+        subprocess_dict['phase'] = float(data.get('phase'))
```

Comment:
`data.get("some key")` would result to `None` if nothing was assigned to that key. I think we may use data["some key"] that way it outputs KeyError whenever key is empty and it would also remove the need for checking whether the keys are present or not.

Reply:
i feel like either way is valid, and there is no need to change for dict[key] since both methods will yield same result with very little performance difference. in fact i would argue that doing it in 2 steps is easier to read and understand

Reply:
the issue is more making sure that `subprocess_dict["character"]` is not an empty character or `None` which might cause us some bugs later in the `stream.py`, or even worse, collect garbage data..